src/DataStream/VectorField.cc

```diff
@@ -185,34 +212,92 @@ void VectorField::FillTileData(CARTA::TileData* tile, int32_t x, int32_t y, int3
         }
+        // std::cout << "Test value FillTiledata = " << array[0] << " compression = " << compression_type << std::endl;
     }
 }
 
-void GetTiles(int image_width, int image_height, int mip, std::vector<Tile>& tiles) {
-    int tile_size_original = TILE_SIZE * mip;
-    int num_tile_columns = ceil((double)image_width / tile_size_original);
-    int num_tile_rows = ceil((double)image_height / tile_size_original);
-    int32_t tile_layer = -1;
-    tiles.resize(num_tile_rows * num_tile_columns);
+VectorField::VectorField() : _stopped(false) {}
 
-    for (int j = 0; j < num_tile_rows; ++j) {
-        for (int i = 0; i < num_tile_columns; ++i) {
-            tiles[j * num_tile_columns + i].x = i;
-            tiles[j * num_tile_columns + i].y = j;
-            tiles[j * num_tile_columns + i].layer = tile_layer;
-        }
+void VectorField::StopCalculations() {
+    // stop flag set first so that new calculations are not started, then wait for mutex to invalidate all the on-going calculations
+    _stopped = true;
+
+    // lock the object and the list of calculators :
+    std::unique_lock lock_calculators(_mutex);
+
+    // Invalidate all on-going calculation to stop them :
+    for (auto calculator : _calculators) {
+        calculator->Invalidate();
     }
 }
 
-CARTA::ImageBounds GetImageBounds(const Tile& tile, int image_width, int image_height, int mip) {
-    int tile_size_original = TILE_SIZE * mip;
-    CARTA::ImageBounds bounds;
-    bounds.set_x_min(std::min(std::max(0, tile.x * tile_size_original), image_width));
-    bounds.set_x_max(std::min(image_width, (tile.x + 1) * tile_size_original));
-    bounds.set_y_min(std::min(std::max(0, tile.y * tile_size_original), image_height));
-    bounds.set_y_max(std::min(image_height, (tile.y + 1) * tile_size_original));
-    return bounds;
+void VectorField::SetVectorOverlayParameters(const CARTA::SetVectorOverlayParameters& parameters) {
+    std::cout << "DEBUG : VectorField::SetVectorOverlayParameters called" << std::endl;
+
+    // FUTURE optimisations may required this check and function Equivalent to be back to avoid re-calculation with the same parameters
+    //    but for now calculation happens everytime it is called
+    //    if( VectorFieldCalculator::Equivalent(message,_vector_field_request_message) ){
+    // requesting the same calculation as before -> no need for this
+    //        return false;
+    //    }
+    _parameters = parameters;
+}
+
+bool VectorField::NewCalculation(const std::function<void(CARTA::VectorOverlayTileData&)>& progress_callback, DimsInfo& dims,
+    bool has_stokes_axis, TileCallback tile_callback, bool stokes_changed /*=false*/, bool z_changed /*=false*/) {
+    // making local copy of the message in case it changes as the calculation goes on:
+    auto parameters = _parameters;
+
+    if (!z_changed && stokes_changed && parameters.stokes_intensity() != VectorFieldCalculator::CURRENT &&
```

Comment:
Here we should call a function on the calculator object (called something like `UsesCurrent`) which internally checks whether any of the three sources is `Source::CURRENT`.